pagespeed/envoy/envoy_cluster_manager.cc

```diff
@@ -0,0 +1,104 @@
+#include <iostream>
+#include <memory>
+#include <random>
+
+#include "envoy/stats/store.h"
+
+#include "external/envoy/include/envoy/event/dispatcher.h"
+#include "external/envoy/source/common/api/api_impl.h"
+#include "external/envoy/source/common/config/utility.h"
+#include "external/envoy/source/common/event/real_time_system.h"
+#include "external/envoy/source/common/init/manager_impl.h"
+#include "external/envoy/source/common/local_info/local_info_impl.h"
+#include "external/envoy/source/common/protobuf/message_validator_impl.h"
+#include "external/envoy/source/common/runtime/runtime_impl.h"
+#include "external/envoy/source/common/singleton/manager_impl.h"
+#include "external/envoy/source/common/stats/allocator_impl.h"
+#include "external/envoy/source/common/stats/thread_local_store.h"
+#include "external/envoy/source/common/thread_local/thread_local_impl.h"
+#include "external/envoy/source/common/upstream/cluster_manager_impl.h"
+#include "external/envoy/source/exe/platform_impl.h"
+#include "external/envoy/source/exe/process_wide.h"
+#include "external/envoy/source/extensions/transport_sockets/tls/context_manager_impl.h"
+#include "external/envoy/source/extensions/transport_sockets/well_known_names.h"
+#include "external/envoy/source/server/config_validation/admin.h"
+#include "external/envoy/source/server/options_impl_platform.h"
+
+namespace net_instaweb {
+
+EnvoyClusterManager::EnvoyClusterManager()
+    : init_watcher_("envoyfetcher", []() {}), secret_manager_(config_tracker_),
+      validation_context_(false, false), init_manager_("init_manager"),
+      local_info_(new Envoy::LocalInfo::LocalInfoImpl(
+          {}, Envoy::Network::Utility::getLocalAddress(Envoy::Network::Address::IpVersion::v4),
+          "envoyfetcher_service_zone", "envoyfetcher_service_cluster",
+          "envoyfetcher_service_node")),
+      stats_allocator_(symbol_table_), store_root_(stats_allocator_),
+      http_context_(store_root_.symbolTable()) {
+  initClusterManager();
+}
+
+EnvoyClusterManager::~EnvoyClusterManager() {
+  tls_.shutdownGlobalThreading();
+  store_root_.shutdownThreading();
+  if (cluster_manager_ != nullptr) {
+    cluster_manager_->shutdown();
+  }
+  tls_.shutdownThread();
+}
+
+void configureComponentLogLevels(spdlog::level::level_enum level) {
+  // TODO(oschaaf): Add options to tweak the log level of the various log tags
+  // that are available.
+  Envoy::Logger::Registry::setLogLevel(level);
+  Envoy::Logger::Logger* logger_to_change = Envoy::Logger::Registry::logger("main");
+  logger_to_change->setLevel(level);
+}
+
+void EnvoyClusterManager::initClusterManager() {
+  auto logging_context = std::make_unique<Envoy::Logger::Context>(spdlog::level::from_str("trace"),
+                                                                  "[%T.%f][%t][%L] %v", log_lock);
+  configureComponentLogLevels(spdlog::level::from_str("trace"));
+
+  api_ = std::make_unique<Envoy::Api::Impl>(platform_impl_.threadFactory(), store_root_,
```

Comment:
It would be nice to add a custom log sink which would forward this to PageSpeed's message handler concept.

Reply:
yes, will implement custom log sink to forward logs to pagespeed message handler.

Reply:
Implemented envoy custom log sink https://github.com/apache/incubator-pagespeed-mod/pull/1969/commits/98691261e660d67e6f31d0030b47426f5f93f801

Reply:
Thanks for doing that! Going forward, it's easier to not further increase scope, it's probably better to work on landing what we have now as this already is quite a large PR. Feel free to punt things to TODO's where needed.